### PR TITLE
[RF] Re-implement `MakeNumpyDataFrame` with no interpreter calls

### PR DESCRIPTION
**Pull request by guitargeek**
Re-implement `MakeNumpyDataFrame` with no interpreter calls, and avoid a
explicit lifetime management of Python objects in the RDNumpyDS helper.

Like this, the former `RNumpyDS.hxx` helper could be moved to
`tree/dataframe`, as it doesn't import `Python.h` anymore.

The `RNumpyDS` also got renamed to `RVecDS`, because it takes RVecs and
not NumPy arrays directly.

### CODE REVIEW

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:197]: Just to try, can we see if `import cppyy` works? In this other PR https://github.com/root-project/root/pull/15011 it was not necessary to go through `ROOT.module` and it would just look better if we could use directly `cppyy` IMHO

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:472]: Can this be made a module-level private function?

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:496]: Also this trick is quite nice!

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:186]: Is this extra function necessary like the deleter function? Or can be just done inline in `RDataFrameAsNumpy`?

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:182]: I guess it's alright to have a global dict, we have a different address for each dictionary of numpy arrays per each process, so different RDataFrame can be created for different numpy sources. Just adding the comment to double check with you

**Review by vepadulano — CHANGES_REQUESTED** [review]
Thank you, this is another nice improvement to our pythonizations! Some first comments for discussion.

One general point regarding the overall design. I wonder if the `RVecDS` data source can be made to have a map/vector of `RVec`s as data member instead of `void *` and those RVecs should be views on the numpy arrays. This would make the design intent clearer and possibly remove the need for a custom deleter that calls into a Python function

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:186]: No, that's not necessary. I will remove it.

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:472]: Sure!

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:496]: Thanks!

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:182]: Yes exactly, that its the idea!

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:197]: Ok, I did that

**Comment by guitargeek** [comment]
> One general point regarding the overall design. I wonder if the `RVecDS` data source can be made to have a map/vector of `RVec`s as data member instead of `void *` 

Thanks for the review! I have changed the `RVecDS` to take a map to `RVecs`, and not pointers to it.

> and those RVecs should be views on the numpy arrays. This would make the design intent clearer and possibly remove the need for a custom deleter that calls into a Python function

This is not possible right now, because RVecs are not true "views on the numpy arrays" in the sense that their existence keeps the numpy array alive. This is only hacked into the Python side by adding an `__adopted__` member to the RVec python proxy:
https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py#L135

As soon as the RVec gets passed to the C++ side we are losing this reference. So we can't work around a custom deleter in the `RVecDS`.

Well, there would be a way, which is to implement a similar mechanism of implementing a deleter that calls to the Python side in RVec itself! But maybe that's not for this PR :slightly_smiling_face:

**Review by vepadulano — APPROVED** [review]
Thank you so much, I think the code looks good!

Last steps before merging:
* Adopt clang-tidy suggestions
* Wait for CI to be green

**Pull request merged**
